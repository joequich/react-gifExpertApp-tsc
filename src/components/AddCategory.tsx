import React, { ChangeEvent, Dispatch, FormEvent, SetStateAction, useState } from 'react';
import PropTypes from 'prop-types';

interface Props {
    setCategories: Dispatch<SetStateAction<string[]>>;
}

export const AddCategory = ({setCategories}: Props) => {
    const [inputValue, setInputValue] = useState<string>('');

    const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e: FormEvent<HTMLElement>) => {
        e.preventDefault();

        if (inputValue.trim().length > 2) {
            setCategories(cats => {
                const exist = cats.find(cat => cat === inputValue);
                if (exist) return [...cats]; 
                return [inputValue, ...cats]
            });
            setInputValue('');
        }
    }

    return (
        <>
            <form role="search" aria-label="Search a gif" onSubmit={handleSubmit}>
                <input type="text" placeholder="Search" value={inputValue} onChange={handleInputChange}/>
            </form>
        </>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}