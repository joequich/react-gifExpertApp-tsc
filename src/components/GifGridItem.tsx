import React from 'react'
import PropTypes from 'prop-types';

interface IGifs {
    id: string;
    title: string;
    url: string;
}

export const GifGridItem = ({id, title, url}:IGifs) => {
    return (
        <div className="card animate__animated animate__fadeInUp">
           <img src={url} alt={title} />
            <p>{title}</p>
        </div>
    )
}

GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
};
GifGridItem.defaultProps = {
    id: ''
};