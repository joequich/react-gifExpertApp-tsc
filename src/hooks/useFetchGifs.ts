import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs";

interface IGifs{
    id: string;
    title: string;
    url: string;
}

export interface IFetchGifs {
    data: IGifs[];
    loading: boolean;
}

export const useFetchGifs = (category:string) => {
    const [state, setState] = useState<IFetchGifs>({
        data: [],
        loading: true
    });

    useEffect(() => {
        getGifs(category)
            .then( imgs => {
                setState({
                    data: imgs,
                    loading: false
                });
            })
    }, [category])

    return state;
}