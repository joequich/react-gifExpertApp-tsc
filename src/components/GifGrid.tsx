import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import GifGridItem from './GifGridItem';
import PropTypes from 'prop-types';

type Props = {
    category: string;
}

interface IGifs {
    id: string;
    title: string;
    url: string;
}

export const GifGrid = ({category}:Props) => {
    const { data: images, loading} = useFetchGifs(category);

    return (
        <>
            <li>
                <h3 className="category-title">{ category }</h3>
                { loading && <p className="animate__pulse">Loading...</p> }
                <div role="list" className="card-grid animate__slideInUp">
                    {images.map( (img:IGifs) => (
                        <GifGridItem
                        key={img.id}
                        {...img}
                        />
                    ))}
                </div>
            </li>
        </>
    )
}

GifGrid.propType = {
    category: PropTypes.string.isRequired
}
