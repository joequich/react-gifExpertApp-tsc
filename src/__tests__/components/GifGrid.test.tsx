import { render, screen } from '@testing-library/react';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs, IFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('GifGrid', () => {

    const category = 'music';
    const mockedUseFetchGifs = useFetchGifs as jest.Mock<IFetchGifs>;

    it('render component correctly', () => {
        mockedUseFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });
        const { asFragment } = render(<GifGrid category={category}/>);
        expect(asFragment()).toMatchSnapshot();
    });

    it('render items when loading images useFetchGifs', () => {
        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/cualquier/coas.png',
            title: 'Cualquier cosa'
        },
        {
            id: '123',
            url: 'https://localhost/cualquier/coas.png',
            title: 'Cualquier cosa'
        }];

        mockedUseFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        render(<GifGrid category={category}/>);
        expect(screen.queryByText('Loading...')).toBeFalsy();
        expect(screen.queryAllByRole('listitem').length).toBe(gifs.length);
    });
});