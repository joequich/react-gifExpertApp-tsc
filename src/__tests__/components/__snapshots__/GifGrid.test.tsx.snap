// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifGrid render component correctly 1`] = `
<DocumentFragment>
  <li>
    <h3
      class="category-title"
    >
      music
    </h3>
    <p
      class="animate__pulse"
    >
      Loading...
    </p>
    <div
      class="card-grid animate__slideInUp"
      role="list"
    />
  </li>
</DocumentFragment>
`;
